File: GuardianMonsters/ui/src/de/limbusdev/guardianmonsters/ui/metamorphosis/MetamorphosisHUD.java
package de.limbusdev.guardianmonsters.ui.metamorphosis;

import com.badlogic.gdx.scenes.scene2d.Action;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.actions.Actions;
import com.badlogic.gdx.scenes.scene2d.ui.Button;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.ImageButton;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.utils.Align;
import com.badlogic.gdx.utils.I18NBundle;

import de.limbusdev.guardianmonsters.assets.paths.AssetPath;
import de.limbusdev.guardianmonsters.media.IMediaManager;
import de.limbusdev.guardianmonsters.scene2d.AnimatedImage;
import de.limbusdev.guardianmonsters.services.Services;
import de.limbusdev.guardianmonsters.ui.AHUD;
import de.limbusdev.guardianmonsters.ui.Constant;
import de.limbusdev.guardianmonsters.ui.widgets.ParticleEffectActor;

/**
 * MetamorphosisHUD
 *
 * @author Georg Eckert 2017
 */

public class MetamorphosisHUD extends AHUD {

    private AnimatedImage animation;
    private Image imgBefore, imgAfter;
    private Button ok;
    private Label label;

    public MetamorphosisHUD(Skin skin, int speciesID, int formerMetaForm, int newMetaForm) {
        super(skin);

        I18NBundle bundle = Services.getL18N().General();
        final IMediaManager media = Services.getMedia();

        String[] monsterNames = {
            Services.getL18N().getLocalizedGuardianName(speciesID, formerMetaForm),
            Services.getL18N().getLocalizedGuardianName(speciesID, newMetaForm)
        };

        final String[] messages =  {
            bundle.format("monster_metamorphs", monsterNames[0]),
            bundle.format("monster_metamorph_complete", monsterNames[0], monsterNames[1])
        };


        // TODO metamorph only at node activation

        // Actor Creation
        Image bg = media.getMetamorphosisBackground();
        animation = media.getMetamorphosisAnimation();
        imgBefore = new Image(media.getMonsterSprite(formerMetaForm,0));
        imgAfter  = new Image(media.getMonsterSprite(newMetaForm,0));
        ok = new ImageButton(skin, "burgund-close");
        label = new Label(messages[0], skin, "burgund");
        label.setSize(420,64);

        ParticleEffectActor particleActor = new ParticleEffectActor("metamorphosis");
        particleActor.setPosition(214,136);


        // Listeners
        ok.addListener(new ClickListener() {
            @Override
            public void clicked(InputEvent event, float x, float y) {
                stage.addAction(Actions.sequence(
                    Services.getAudio().getMuteAudioAction(AssetPath.Audio.Music.METAMORPHOSIS),
                    Actions.fadeOut(1),
                    Actions.delay(1),
                    Actions.run(() -> Services.getScreenManager().popScreen())
                ));
            }
        });

        // Layout
        bg.setPosition(0,0,Align.bottomLeft);
        animation.setPosition(Constant.RES_X/2-128, Constant.RES_Y/2-128+30, Align.bottomLeft);
        imgBefore.setPosition(Constant.RES_X/2-64,Constant.RES_Y/2-34,Align.bottomLeft);
        imgAfter.setPosition(Constant.RES_X/2-64,Constant.RES_Y/2-34,Align.bottomLeft);
        ok.setPosition(Constant.RES_X-16,18,Align.bottomRight);
        label.setPosition(4,4,Align.bottomLeft);

        Action metaAction = Actions.sequence(
            Actions.delay(4),
            Services.getAudio().getMuteAudioAction(AssetPath.Audio.Music.METAMORPHOSIS),
            Actions.delay(1),
            Actions.run(() -> {
                label.remove();
                stage.addActor(animation);
                Services.getAudio().playSound(AssetPath.Audio.SFX.METAMORPHOSIS);
                stage.addActor(label);
            }),
            Actions.delay(2),
            Actions.run(() -> {
                animation.remove();
                imgBefore.remove();
                label.remove();
                stage.addActor(imgAfter);
                stage.addActor(animation);
                stage.addActor(label);
            }),
            Actions.delay(3),
            Actions.run(() -> {
                Services.getAudio().playMusic(AssetPath.Audio.Music.VICTORY_FANFARE);
                label.setText(messages[1]);
            }),
            Actions.delay(5.5f),
            Services.getAudio().getFadeInMusicAction(AssetPath.Audio.Music.METAMORPHOSIS),
            Actions.run(() -> stage.addActor(ok))
        );

        // Adding actors to stage
        stage.addActor(bg);
        stage.addActor(particleActor);
        stage.addActor(imgBefore);
        stage.addAction(metaAction);
        stage.addActor(label);

        particleActor.start();
    }

    @Override
    protected void reset() {

    }

    @Override
    public void show() {
        Services.getAudio().playLoopMusic(AssetPath.Audio.Music.METAMORPHOSIS);
    }

    @Override
    public void hide() {
        super.hide();
        Services.getAudio().stopMusic(AssetPath.Audio.Music.METAMORPHOSIS);
    }
}

File: GuardianMonsters/ui/src/de/limbusdev/guardianmonsters/ui/metamorphosis/MetamorphosisHUD.kt
package de.limbusdev.guardianmonsters.ui.metamorphosis

import com.badlogic.gdx.scenes.scene2d.Action
import com.badlogic.gdx.scenes.scene2d.actions.Actions.*
import com.badlogic.gdx.scenes.scene2d.ui.*
import com.badlogic.gdx.utils.Align
import de.limbusdev.guardianmonsters.assets.paths.AssetPath
import de.limbusdev.guardianmonsters.scene2d.AnimatedImage
import de.limbusdev.guardianmonsters.scene2d.replaceOnClick
import de.limbusdev.guardianmonsters.services.Services
import de.limbusdev.guardianmonsters.ui.AHUD
import de.limbusdev.guardianmonsters.ui.Constant
import de.limbusdev.guardianmonsters.ui.widgets.ParticleEffectActor
import ktx.actors.*
import com.badlogic.gdx.scenes.scene2d.actions.Actions.run as runAction

/**
 * MetamorphosisHUD
 *
 * @author Georg Eckert 2017
 */

class MetamorphosisHUD
(
        skin: Skin,
        speciesID: Int,
        formerMetaForm: Int,
        newMetaForm: Int
)
    : AHUD(skin)
{
    private val animation: AnimatedImage
    private val imgBefore: Image
    private val imgAfter: Image
    private val okButton: Button
    private val label: Label

    init
    {
        val bundle = Services.I18N().General()
        val media = Services.Media()

        val monsterNames = arrayOf(

                Services.I18N().getLocalizedGuardianName(speciesID, formerMetaForm),
                Services.I18N().getLocalizedGuardianName(speciesID, newMetaForm)
        )

        val messages = arrayOf(

                bundle.format("monster_metamorphs", monsterNames[0]),
                bundle.format("monster_metamorph_complete", monsterNames[0], monsterNames[1])
        )

        // Actor Creation
        val background  = media.getMetamorphosisBackground()
        animation       = media.getMetamorphosisAnimation()
        imgBefore       = Image(media.getMonsterSprite(speciesID, formerMetaForm))
        imgAfter        = Image(media.getMonsterSprite(speciesID, newMetaForm))
        okButton        = ImageButton(skin, "burgund-close")
        label           = Label(messages[0], skin, "burgund")
        label.setSize(labelWidth, labelHeight)

        val particleActor = ParticleEffectActor("metamorphosis")
        particleActor.setPosition(particlesX, particlesY)

        // Listeners
        okButton.replaceOnClick()
        {
            val sequence = muteAudio() then fadeOut(.5f) then popScreen()
            stage+=sequence
        }

        // Layout
        background  .setPosition(bgX, bgY, bgAlign)
        animation   .setPosition(animationX, animationY, animationAlign)
        imgBefore   .setPosition(startImgX, startImgY, startImgAlign)
        imgAfter    .setPosition(afterImgX, afterImgY, afterImgAlign)
        okButton    .setPosition(okButtonX, okButtonY, okButtonAlign)
        label       .setPosition(labelX, labelY, labelAlign)

        val showAnimation = runAction()
        {
            label.remove()
            stage+=animation
            playMetamorphosisSFX()
            stage+=label
        }
        val removeAnimation = runAction()
        {
            animation.remove()
            imgBefore.remove()
            label.remove()
            stage+=imgAfter
            stage+=animation
            stage+=label
        }
        val playVictorySFX = runAction()
        {
            playVictorySFX()
            label.txt = messages[1]
        }
        val addOkButton = runAction { stage+=okButton }

        val metaAction =
                delay(4f)       then
                muteAudio()     then
                delay(1f)       then
                showAnimation   then
                delay(2f)       then
                removeAnimation then
                delay(3f)       then
                playVictorySFX  then
                delay(5.5f)     then
                fadeInMusic()   then
                addOkButton

        // Adding actors to stage
        stage+=background
        stage+=particleActor
        stage+=imgBefore
        stage+=metaAction
        stage+=label

        particleActor.start()
    }

    override fun reset() {}

    override fun show()
    {
        startBGMusic()
    }

    override fun hide()
    {
        super.hide()
        stopBGMusic()
    }

    companion object
    {
        const val particlesX = 214f
        const val particlesY = 136f

        const val bgX = 0f
        const val bgY = 0f
        const val bgAlign = Align.bottomLeft

        const val animationX = Constant.RES_X/2f - 128
        const val animationY = Constant.RES_Y/2f - 128 + 30
        const val animationAlign = Align.bottomLeft

        const val startImgX = Constant.RES_X/2f - 64
        const val startImgY = Constant.RES_Y/2f - 34
        const val startImgAlign = Align.bottomLeft

        const val afterImgX = Constant.RES_X/2f - 64
        const val afterImgY = Constant.RES_Y/2f - 34
        const val afterImgAlign = Align.bottomLeft

        const val okButtonX = Constant.RES_X - 16f
        const val okButtonY = 18f
        const val okButtonAlign = Align.bottomRight

        const val labelX = 4f
        const val labelY = 4f
        const val labelAlign = Align.bottomLeft
        const val labelWidth = 420f
        const val labelHeight = 64f

        fun startBGMusic()          = Services.Audio().playLoopMusic(AssetPath.Audio.Music.METAMORPHOSIS)
        fun stopBGMusic()           = Services.Audio().stopMusic(AssetPath.Audio.Music.METAMORPHOSIS)
        fun playVictorySFX()        = Services.Audio().playMusic(AssetPath.Audio.Music.VICTORY_FANFARE)
        fun playMetamorphosisSFX()  = Services.Audio().playSound(AssetPath.Audio.SFX.METAMORPHOSIS)

        fun muteAudio(): Action     = Services.Audio().createMuteAction(AssetPath.Audio.Music.METAMORPHOSIS)
        fun fadeInMusic(): Action   = Services.Audio().getFadeInMusicAction(AssetPath.Audio.Music.METAMORPHOSIS)
        fun popScreen(): Action     = runAction { Services.ScreenManager().popScreen() }
    }
}
